doc/gyrid.1
.TH "gyrid" 1
.SH NAME
gyrid \- A mobile device scanner and logger.
.SH SYNOPSIS
.B /etc/init.d/gyrid
start|stop|restart
.SH DESCRIPTION
Gyrid continuously scans the neighborhood for Bluetooth-enabled devices which are publicly visible, given there is a Bluetooth adapter connected. It can scan with multiple adapters simultaneously too.

The hardware id's of the discovered devices are saved to a logfile including a timestamp and their device class (which is a number that indicates whether the device is a phone, a carkit, etc.). Optionally, the RSSI value of each detection can be saved as well.

When there is a WiFi adapter attached which is not managed by NetworkManager, it is used to monitor the radiospectrum for WiFi activity. All received packets are logged, with respect to the users privacy. Only the 802.11 packet headers are saved, the packet content is not.

The idea behind this is scientifical. When multiple of these scanners are placed at known positions, the logfiles can be combined into a (geo)database and used for further analysis. This can be done in realtime via a network server or in postprocessing.
.SH OPTIONS
.TP
.B start
Start Gyrid.
.TP
.B stop
Stop Gyrid.
.TP
.B restart
Restart Gyrid.
.SH DEBUG
.B /usr/share/gyrid/gyrid-start debug [--no-log]

To start Gyrid in debug mode, use the gyrid-start script. Debug mode does not daemonize the process and can be ended by issuing Ctrl+C on the console.

Debug mode is more verbose, printing all activity to standard output. This might be useful to spot bugs or issues more easily.

With the optional --no-log flag set, no logs are written to disk.
.SH FILES
.TP
/etc/gyrid/gyrid.conf
Configuration file where a few options can be specified. The default configuration file documents all options.
.TP
/etc/gyrid/blacklist.conf
File where blacklisted MAC-addresses can be listed, which will be exluded from registration. One can list either a full address or only the starting part.
In the latter case, all addresses starting with that part will be blacklisted.
.TP
/etc/gyrid/arduino.conf
File where the Arduino-based rotating antenna configuration resides. It is a CSV file with in the first column the MAC-address of the Bluetooth sensor, in the second column the Arduino id (f.ex. arduino001) and in the third column the turning resolution (i.e. in how many parts the 180 degree arc is divided).
.TP
/var/log/gyrid/
This is the directory where all logfiles are stored.
.SH BUGS
Bugs and patches are welcome via e-mail at <roel.huybrechts@ugent.be>
.SH AUTHOR
Roel Huybrechts <roel.huybrechts@ugent.be>, Ghent University


ChangeLog
gyrid 0.8.3

    * FIX: Locking in AckMap.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Wed, 15 May 2013 15:15:56 +0200

gyrid 0.8.2

    * FIX: Force hashing of WiFi SSID's.
    * UPD: Terminology and documentation.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Wed, 15 May 2013 09:58:43 +0200

gyrid 0.8.1

    * FIX: MANIFEST: include core.py

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Tue, 14 May 2013 15:52:00 +0200

gyrid 0.8

    * ADD: WiFi scanning.
    * ADD: Hashing of privacy sensitive data.
    * UPD: Replace network protocol by a binary version.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Tue, 14 May 2013 13:10:35 +0200

gyrid 0.7

    * ADD: Inquiry logging.
    * MERGE: Support for Arduino-based rotating platforms.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Wed, 13 Mar 2013 11:21:27 +0100

gyrid 0.6.2-1

    * ADD: Differentiate between live and cached data when sending data
             over the network.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Wed, 17 Oct 2012 13:11:50 +0200

gyrid 0.6.2

    * FIX: Make init script more LSB compliant.
    * ADD: Resend cached items when they're not ack'ed after a certain time.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Wed, 17 Oct 2012 13:03:37 +0200

gyrid 0.6.1-1

    * FIX: Make init script more LSB compliant.
    * ADD: Blacklist to exclude certain MAC-addresses from registration.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Sun, 07 Oct 2012 15:39:35 +0200

gyrid 0.6.1

    * FIX: Depend on bluetooth subsystem in init script.
    * FIX: Write a new default configuration when current one fails to parse.
    * ADD: Option to configure network caching limit.
    * UPD: Switch to LSB init functions in init script.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Fri, 05 Oct 2012 14:41:22 +0200

gyrid 0.6

    * FIX: Remove obsolete daemon code and switch to start-stop-daemon.
             This should work with more recent Python versions (2.5+) too.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Wed, 03 Oct 2012 10:40:54 +0200

gyrid 0.5.1.1

    * ADD: Support for Arduino rotating antenna platform.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Wed, 13 Feb 2013 12:21:11 +0100

gyrid 0.5.1

    * FIX: Only die on SSL errors if it are actually SSL errors.
    * UPD: Allow network caching up to 250 MB instead of just 25.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Sat, 30 Jul 2011 16:26:39 +0200

gyrid 0.5

    * FIX: Use millisecond resolution timestamps in network module.
    * UPD: IEEE OUI data update.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Mon, 11 Jul 2011 21:34:57 +0200

gyrid 0.4.9

    * ADD: Playback discoverer for testing purposes.
    * UPD: Include testing package in release.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Wed, 06 Jul 2011 11:59:07 +0200

gyrid 0.4.8

    * FIX: Net: add some checks before trying to send lines.
    * ADD: Net: sensor started/stopped scanning status messages.
    * ADD: Net: uptime status messages.
    * ADD: Net: Gyrid daemon connected/disconnected status messages.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Wed, 18 May 2011 21:39:03 +0200

gyrid 0.4.7

    * FIX: Net: restart middleware when not running.
    * ADD: Net: inquiry state notification.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Sat, 14 May 2011 18:00:33 +0200

gyrid 0.4.6

    * FIX: Start Gyrid earlier in boot process.
    * FIX: Turn off LED's on logger shutdown.
    * ADD: Net: client/server keepalive.
    * ADD: Net: client-side buffering and caching.
    * ADD: Net: documentation.
    * UPD: OUI data.
    * UPD: Clean out this changelog.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Tue, 26 Apr 2011 16:25:51 +0200

gyrid 0.4.5

    * FIX: Initialise DBus threading on startup.
    * FIX: Save loggers between discoveries.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Wed, 30 Mar 2011 11:19:02 +0200

gyrid 0.4.4

    * ADD: option to disable SSL on client-server connection

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Thu, 03 Mar 2011 16:18:20 +0100

gyrid 0.4.3

    * FIX: time difference function in zippingfilehandler
    * FIX: record non-RSSI inquiries on adapter that don't support RSSI
           inquiry mode
    * ADD: option to temporarily disable LED interaction

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Tue, 15 Feb 2011 19:06:54 +0100

gyrid 0.4.2

    * ADD: option to record only detections with a RSSI above
           a certain threshold

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Mon, 15 Nov 2010 19:37:52 +0100

gyrid 0.4.1

    * FIX: strip colons from directory names
    * FIX: stop logging while in debug mode
    * FIX: net: strip colons from MAC-addresses
    * FIX: make ALIX LED switching more efficient
    * ADD: net: make sensor's MAC optional
    * ADD: net: make timestamp optional
    * UPD: manpage

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Thu, 11 Nov 2010 16:34:40 +0100

gyrid 0.4

    * ADD: networking support, talking to a remote server either on the LAN
           or via the internet.
    * REM: track mode, use a combination of tail and grep on the logfile
           instead
    * REM: Bluetooth reporting functionality, it has never really worked
    * FIX: logging while device pool is locked
    * FIX: error in writing inquiry mode on Bluetooth 2.1 devices

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Wed, 27 Oct 2010 11:09:29 +0200

gyrid 0.3.3

    * FIX: excluded_devices option so that it detects 0 as a valid
           value.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Sun, 27 Jun 2010 20:25:08 +0200

gyrid 0.3.2

    * FIX: Catch exceptions raised inside threads.
    * FIX: excluded_devices option.
    * UPD: Hide interacting_devices option. Sending reports via Bluetooth
           is currently untested and unsupported.

-- Roel Huybrechts <roel.huybrechts@ugent.be>  Thu, 24 Jun 2010 21:26:58 +0200

gyrid 0.3.1

    * ADD: Option to exclude certain adapters from scanning,
           based on their HCI device ID.

gyrid/gyrid.py
#-*- coding: utf-8 -*-
#
# This file belongs to Gyrid.
#
# Gyrid is a mobile device scanner.
# Copyright (C) 2009-2012  Roel Huybrechts
#
# This program is free software: you can redistribute it and/or modify
# it under the terms of the GNU General Public License as published by
# the Free Software Foundation, either version 3 of the License, or
# (at your option) any later version.
#
# This program is distributed in the hope that it will be useful,
# but WITHOUT ANY WARRANTY; without even the implied warranty of
# MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
# GNU General Public License for more details.
#
# You should have received a copy of the GNU General Public License
# along with this program.  If not, see <http://www.gnu.org/licenses/>.

import gobject
import os
import re
import signal
import sys
import time
import threading
import traceback

import scanmanager

class Main(object):
    """
    Main class of the Bluetooth tracker.
    """
    def __init__(self, configfile, errorlogfile):
        """
        Initialisation of configuration, logging and DBus connection.

        @param  configfile      URL of the configfile.
        @param  errorlogfile    URL of the errorlogfile.
        """
        self.errorlogfile = errorlogfile
        self.errorlog = open(self.errorlogfile, 'a')

        self._install_thread_excepthook()
        sys.excepthook = self._handle_exception
        signal.signal(signal.SIGTERM, self._catch_sigterm)
        self.stopping = False

        self.configfile = configfile
        self.mgr = scanmanager.DefaultScanManager(self)

        self.main_loop = gobject.MainLoop()
        gobject.threads_init()

    def pass_args(self, *args):
        """
        Handle commandline arguments passed to Gyrid.

        @param  *args    sys.argv arguments.
        """
        argerr = False
        self.debug_mode = False
        if len(args) == 2:
            if args[1] == 'debug':
                self.debug_mode = True
                self.mgr.set_debug_mode(True, False)
            else:
                argerr = True
        elif len(args) == 3 and args[1] == 'debug' and args[2] == '--no-log':
                self.debug_mode = True
                self.mgr.set_debug_mode(True, True)
        elif len(args) > 1:
            argerr = True

        if argerr:
            self.log_error('Error', 'Wrong set of arguments %s' % str(args))
            sys.stderr.write('Gyrid: Error: Wrong set of arguments.\n')
            sys.exit(2)
        else:
            self.run()

    def _handle_exception(self, etype, evalue, etraceback):
        """
        Handle the exception by writing information to the error log.
        """
        exc = ''.join(traceback.format_exception(etype, evalue, etraceback))
        self.log_error('Error', exc)
        sys.stderr.write("Error: unhandled exception: %s, %s\n\n" % \
            (etype.__name__, evalue))
        sys.stderr.write(' '.join(traceback.format_exception(etype, evalue,
            etraceback)))

    def _install_thread_excepthook(self):
        """
        Workaround for sys.excepthook thread bug
        (http://bugs.python.org/issue1230540)

        From
        http://spyced.blogspot.com/2007/06/workaround-for-sysexcepthook-bug.html
        """
        init_old = threading.Thread.__init__
        def init(self, *args, **kwargs):
            init_old(self, *args, **kwargs)
            run_old = self.run
            def run_with_except_hook(*args, **kw):
                try:
                    run_old(*args, **kw)
                except (KeyboardInterrupt, SystemExit):
                    raise
                except:
                    sys.excepthook(*sys.exc_info())
            self.run = run_with_except_hook
        threading.Thread.__init__ = init

    def log_error(self, level, message):
        """
        Write the given message to the error log.

        @param  level    The level of the message, f. ex. Warning or Error
        @param  message  The message to write.
        """
        self.errorlog.write("%(tijd)s %(level)s: %(message)s\n" % \
            {'tijd': time.strftime('%Y%m%d-%H%M%S-%Z'),
             'level': level, 'message': message})
        self.errorlog.flush()

    def _catch_sigterm(self, signum, frame):
        """
        Called when a SIGTERM signal is received, ie. when the process is
        killed. Shutdown the manager and exit.
        """
        self.stopping = True
        self.mgr.stop()
        sys.exit(0)

    def run(self):
        """
        Connect the AdapterAdded signal (DBus/BlueZ) to its handler and
        start the Bluetooth discoverer.
        """
        debugstr = " in debug mode" if self.debug_mode else ""
        self.mgr.log_info("Started" + debugstr)

        self.mgr.init()

        try:
            self.main_loop.run()
        except KeyboardInterrupt:
            self.stop()

    def stop(self):
        """
        Stop discovering, stop the logger and cleanly close the logfile.
        """
        self.stopping = True
        self.mgr.log_info("Shutting down")
        self.mgr.stop()
        sys.exit(0)
